Assignment3/PointAndGrid.py

```diff
@@ -0,0 +1,260 @@
+                          (row+1, col-1),(row+1, col),(row, col+1),(row+1, col+1)]
+        for coord in possible_coord:
+            if self._validCoordinates(coord):
+                result.append(Point(self, coord, self.getElement(coord)))
+        return result
+    
+    def ifWordInGrid(self, word):
+        """
+        Return True if the word can be formed in grid, False otherwise.
+        :param word: str
+        :return: bool
+        """
+        if self.findString(word) is None:
+            return False
+        return True
+    
+    ############################## THE SOLUTION TO ASSIGNMENT 3 ##########################################
+
+    def findString(self, word):
+        """
+        Find the word in the grid where we can starting from one position move to any neighbors,
+        bub can't visit the same cell twice. Return the path if exists, None otherwise.
+        :param word: string
+        :return: bool
+        """
+        startPoints = self.findElementInGrid(word[0])
+        path = []
+        for startP in startPoints: # we go further only if the first letter of word is in grid
+            path = [startP]
+            path = self._findPathRecursively(startP, word, path, [])
+        if self._pathIsWord(path, word):
+            return path
+
+    def _pathIsWord(self, path, word):
+        return [p.value for p in path] == list(word)
+
+    def _findPathRecursively(self, point, word, path, visited):
+        """
+        Helper-function to find a word in a grid. Finds the longest path to form the word (may be only a part)
+        :param point: Point
+        :param word: str
+        :param path: list(Point)
+        :param visited: list(Point)
+        :return: None
+        """
+        if self._pathIsWord(path, word):
+            return path
+
+        possible_ways = self.findPossibleWays(path, point, word, visited)
+        if possible_ways == []:
+            visited.append(path.pop(-1)) # we at the same time pop the last element from path and append it to list visited.
+            if path:
+                self._findPathRecursively(path[-1], word, path, visited)
+
+        for neighbor in possible_ways:
+            if neighbor not in path and neighbor not in visited:
+                path.append(neighbor)
+                self._findPathRecursively(neighbor, word, path, visited)
+                
+        return path
+
+    def findPossibleWays(self, path, point, word, visited):
+        """
+        Return the list of Points that can be the next element in the word.
+        :return: list(Point)
+        """
+        neighbors = self.findNeighborsOfPoint(point.coordinates)
+        ways = []
+        for n in neighbors:
+            if self._pathIsPartOfWord(path + [n], word) and n not in path: # we can't use one Point twice
+                if not visited or n != visited[-1]: # it is of no use to go in the Point from where we just returned

```

Comment:
In fact, it is forbidden to re-visit any past points, so this check should be even more strict.